aip/0162.md

```diff
@@ -0,0 +1,229 @@
+  resource for reference later.
+
+## Guidance
+
+APIs **may** store a revision history for a resource if it is useful to users.
+
+APIs implementing resources with a revision history **must** provide a
+`revision_id` field on the resource:
+
+```proto
+message Book {
+  // The name of the book.
+  string name = 1;
+
+  // Other fields…
+
+  // The revision ID of the book.
+  // This is automatically changed whenever the contents of the book
+  // change in any way.
+  string revision_id = 99 [(google.api.field_behavior) = OUTPUT_ONLY];
+}
+```
+
+- The `revision_id` field **should** be a string and contain a short,
+  automatically-generated hash. A good rule of thumb is the last eight
+  characters of a UUID4.
+- The field **must** document when new revisions are created.
+
+### Referencing revisions
+
+When it is necessary to refer to a specific revision of a resource, APIs
+**must** use the following syntax: `{resource_name}@{revision_id}`. For
+example:
+
+    publishers/123/books/les-miserables@c7cfa2a8
+
+APIs **should** generally accept a resource reference at a particular revision
+in any place where they ordinarily accept the resource name, except in
+situations that mutate the resource.
+
+**Important:** APIs **must not** require a revision ID, and **must** default to
+the current revision if one is not provided, except in methods specifically
+dealing with the revision history (such as rollback) where failing to require
+it would not make sense.
+
+### Getting a revision
+
+APIs implementing resource revisions **should** accept a resource name with a
+revision ID in the standard `Get` method ([AIP-131][]):
+
+```proto
+message GetBookRequest {
+  // The name of the book.
+  //   Example: publishers/123/books/les-miserables
+  //
+  // In order to retrieve a previous revision of the book, also provide
+  // the revision ID.
+  //   Example: publishers/123/books/les-miserables@c7cfa2a8
+  string name = 1;
+}
+```
+
+If the user passes a revision ID that does not exist, the API **must** fail
+with a `NOT_FOUND` error.
+
+### Listing revisions
+
+APIs implementing resource revisions **should** provide a custom method for
+listing the revision history for a resource, with a structure similar to
+standard `List` methods ([AIP-132][]):
+
+```proto
+rpc ListBookRevisions(ListBookRevisionsRequest)
+    returns (ListBookRevisionsResponse) {
+  option (google.api.http) = {
+    get: "/v1/{name=publishers/*/books/*}:listRevisions"
+  }
+}
+```
+
+```proto
+message ListBookRevisionsRequest {
+  // The name of the book to list revisions for.
+  string name = 1;
+
+  // The maximum number of revisions to return per page.
+  int32 page_size = 2;
+
+  // The page token, received from a previous ListBookRevisions call.
+  // Provide this to retrieve the subsequent page.
+  string page_token = 3;
+}
+
+message ListBookRevisionsResponse {
+  // The revisions of the book.
+  repeated BookRevision book_revisions = 1;
```

Comment:
Another way to do this would be `repeated Book book` which in many ways I prefer, but that makes reasoning about "when the revision was current" tougher, and makes revision metadata tougher also.

Reply:
I would much prefer repeated Book book, and for any versioned resource require it to also contain revision_create_time field. Let's shave off one level of indirection.